db/db_impl/db_impl_open.cc

```diff
@@ -1297,10 +1314,12 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
         edit_lists.back().push_back(wal_deletion.get());
       }
 
-      // write MANIFEST with update
-      status = versions_->LogAndApply(cfds, cf_opts, edit_lists, &mutex_,
-                                      directories_.GetDbDir(),
-                                      /*new_descriptor_log=*/true);
+      if (flushed || total_entries > 0) {
```

Comment:
It is needed to avoid the weird case where leader produces a manifest write while in recovery.

Reply:
Hmm, not sure I understand what you mean by that

Reply:
Let's say you are recovering from a MANIFEST snapshot and a log (which contains manifest writes that you need to apply to your snapshot). You open a database and during open RocksDB writes to MANIFEST, before you had a chance to apply the pending manifest writes. Now you have a new MANIFEST branch. :( We can tolerate this on the leader by poisoning the epoch and forcing the followers to recover from snapshot, but that's not great, so we try to avoid it.

Reply:
Looks like this was fixed upstream, from https://github.com/facebook/rocksdb/releases/tag/v7.2.2:
> Fixed a bug in RocksDB DB::Open() which may creates and writes to two new MANIFEST files even before recovery succeeds. Now writes to MANIFEST are persisted only after recovery is successful.

Reply:
Ah I see, thanks for explaining. Are we going to try to update and rebase on the upstream change or just modify here for now? Just curious how we handle these situations after forking from external dependencies

Reply:
I want to let Facebook bake the new RocksDB version for a while before we upgrade.